### ISSUE TITLE
suggestion: autocommit - disable at all or at least to make it disabled by default

### ISSUE THREAD

**Issue by pvanek** [dbeaver/dbeaver#2738]
hi dbeavers,

this tool is pretty cool. Please consider changing connection "autocommit" feature from default "on" to "off". From my experience the autocommit=on causes real issues in all possible working flows leading into accidentally deleted/updated/inserted data.

So chaning on->off can make our lives safer a bit.

(Personally I'd like to see autocommit completely hidden or even removed at all ;))

**Comment by AndrewKhitrin** [comment]
I agree. Autocommit should be "off" by default because it safer

**Comment by serge-rider** [comment]
This is very controversial.
People use DBeaver for too different things. I know that a lot of people never use transactions. And a lot of people don't even understand how transactions work.

I personally often work in auto-commit mode on dev databases. Even if I'll loose some data it is not a big problem.
For production/staging/qa there are connection types (where auto-commit is off by default).

So, I'm not sure about this. But you could try to convince me :)

Or we could make some kind of global profiles (e.g. "SQL developer", "DBA", "Content editor", etc) which will define default settings..

**Comment by pvanek** [comment]
What about to intoduce a dedicated page in the new connection wizard? It would be completely dedicated to autocommit displaying info text about dangers of ac (and pending/locked transactions too).

We have discovered dbeaver in our office today and almost all my devs used "develop" profile for all connections without any doubt. For eveything. It was very scarry, believe me ;)

Or maybe a "first launch wizard" asking for it.

**Comment by benyanke** [comment]
I think there's sane defaults as is - production uses no autocommit, and dev does. I find that most useful for both me and other devs at my shop.

**Comment by amuelz** [comment]
I vote for autocommit = off as the default settings for all database connections (irrespective of production / test / development connection). Working with auto commit on databases is like programming without a software versioning system.

**Comment by siick** [comment]
Same here, i think it's better to have autocommit = off as the default settings for all database connections. And if we want to have autocommit = on by default, then add a parameter to change it. Like that you have the default autocommit  you want.

**Comment by EimanMarzouk** [comment]
I agree, having the option to choose the default setting for autocommit per database connection is the best option

**Comment by p91paul** [comment]
to everybody: autocommit is already a setting that can be changed per connection. Also for groups of connections depending on connection type.
This ticket is here because the default setting for a new connection type is development, and default autocommit for development is on.

I'd argue that current connection wizard no longer offers the explicit option to choose connection type: you have to know which button to press to get to a page where autocommit behavior can be changed. So maybe this is worth improving; that said, autocommit on development connections makes sense.

**Comment by serge-rider** [comment]
@p91paul thanks for explaining this 👍 

Probably we could add "Connection Type" selector somewhere in the connection wizard to make it more obvious.
E.g. on driver selector page near to Project selector. Or right on connection properties page near to Edit Driver button.

WDYT?

**Comment by pvanek** [comment]
To be honest, the autocommit=on (even in DEV env) is a Bad idea (tm). At least a big fat warning should be there: "selecting dev sets autocommit on, you have been warned" or something

**Comment by serge-rider** [comment]
@pvanek Generally yes, but please check my comment https://github.com/dbeaver/dbeaver/issues/2738#issuecomment-355627484
Disabled auto-commit confuses some people as they don't understand how to "really" save data in the database.

I still think that having this option in connection wizard should help more experienced people and solve the problem..

**Comment by p91paul** [comment]
@serge-rider I personally prefer the second option, in the connection property page.

@pvanek if autocommit on dev is a bad for you, you can disable it globally in preferences. By default, autocommit is fine: if you think about it, people with weak database knowledge expect that when they modify a table from the table editor and click save, changes are persisted. Which is not true with autocommit off, and leads to errors (there are plenty of issues here like "I saved it, then after closing and reopening DBeaver it disappeared").
People like you, who are transaction-aware, can also handle editing an option.

**Comment by edmiller3** [comment]
For what it is worth, I just want both sides to understand how much chaos _either_ default value for the setting can cause.  

- When you **disable** auto-commit, every query or update you initiate could potentially open a transaction that will lock tables.  With less knowledgeable analysts running just SELECT queries against your system, they might end up leaving a window open that is locking a table that will cause your overnight processes to fail when they try to update that table.
- When you **enable** auto-commit, if you are performing updates and accidentally make a mistake (oops, I did a DELETE * FROM but forgot to include the WHERE clause!) you can't rollback the transaction.

My point is simply that there is no single "correct" answer to this.  I think that @serge-rider nailed this by placing global default preferences in Database > Connection Types (Production/Development/Test), plus allowing the override on each connection itself.

**Comment by dpsspatial** [comment]
Not sure if something has changed in recent versions, but we're having major issues with transactions we *think* are going through in Auto mode not actually going through and not displaying as actually waiting for manual commit. This is frustrating because there is no notification until you override the Auto button to see what is actually waiting for commit. We're going to provide more details.

**Comment by bast1aan** [comment]
In my opinion opening a transaction for only SELECT queries is quite useless and obstructing, especially if it is not visible there is an open transaction.  SELECT queries is what I do 90% of the time. Because many other tools for PostgreSQL have default autocommit=on, I always explicitly use START TRANSACTION; ... COMMIT; in my queries if I want to use transactions when I do DELETE/UPDATE/INSERT.

Could be I don't oversee things enough, but some feature like this would be great:
- Only start a transaction on DELETE/UPDATE/INSERT/... initiated either with SQL scripts or when editing a data tab
- do not start a transaction for single SELECT queries
- Make it more clear/visible once a transaction is started

**Comment by WicloW** [comment]
Hi, i noticed these last times that my Auto commit is off and once i do one commit it is switching to Auto commit enabled instead to remain in manual commit...
Could it be possible to remain in manual commit?
Cause it is really dangerous on Production DB to be in autocommit.
And checking the autocommit state after each manual commit is not valuable.

**Comment by spardue** [comment]
Wow, I can't believe autocommit is on by default. This is an **extremely bad idea** and makes me doubt the overall design of this application.

**Comment by janbuchar** [comment]
I set my connection type to production without knowing that it disables autocommit. I did a couple of updates which locked various rows and caused parts of the production system to get stuck. In my opinion, this actually defeats the purpose of the "production" connection type, especially for users coming from other tools (adminer/pgadmin in my case) that use autocommit.

**Comment by Diego-Hernandez-Moodys** [comment]
You can prompt me when I try to close the application but not when I attempt to commit a transaction? Smart Commit should be the default at the very least. Committing transactions are only applicable to Updates, Deletes, etc., so if someone is confused about how transactions work, then they should not be performing them.

**Comment by sqlpro20** [comment]
This whole feature is weird. I can understand that the auto commit setting is done per connection type. However, there is no option to set auto commit permanently in connection settings (at least I can't find none). As a result, I need to set each connection I want auto commit permanently disabled as production(!). So either I did not understand something or the Interface is too complicated.

**Comment by p91paul** [comment]
Here is some things you can do:

* set autocommit default for connection type (i.e. you can set development connections as manual commit as well)
![immagine](https://user-images.githubusercontent.com/2157006/208442429-332a82b3-6ae9-44b2-96a9-3cd9469acb48.png)
* set or unset autocommit per connection, ignoring the connection type default (to access this view, right click on a connection in Database Navigator and select "Edit Connection (F4)"
![immagine](https://user-images.githubusercontent.com/2157006/208442715-f4f3c1f9-3b0e-4133-949a-3e6551217f54.png)

**Comment by sqlpro20** [comment]
Thanks, I would have never looked in initialization if a menu point transactions is existing.

**Comment by davestokes** [comment]
Will not change this to a default.
